sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/PagedIterable.java

```diff
@@ -0,0 +1,206 @@
+                }
+
+                return currentPage.next();
+            }
+
+            @Override
+            void addPage(PagedResponse<T> page) {
+                Iterator<T> pageValues = page.getValue().iterator();
+                if (pageValues.hasNext()) {
+                    this.pages.add(pageValues);
+                }
+            }
+        };
+    }
+
+    private Iterable<PagedResponse<T>> iterableByPageInternal() {
+        return () -> new PagedByIterator<T, PagedResponse<T>>(pageRetriever) {
+
+            private final Queue<PagedResponse<T>> pages = new ConcurrentLinkedQueue<>();
+
+            @Override
+            boolean needToRequestPage() {
+                return pages.peek() == null;
+            }
+
+            @Override
+            boolean isNextAvailable() {
+                return pages.peek() != null;
+            }
+
+            @Override
+            PagedResponse<T> getNext() {
+                return pages.poll();
+            }
+
+            @Override
+            void addPage(PagedResponse<T> page) {
+                this.pages.add(page);
+            }
+        };
+    }
+
+    private abstract static class PagedByIterator<T, E> implements Iterator<E> {
+        private static final ClientLogger LOGGER = new ClientLogger(PagedByIterator.class);
+
+        private final Function<String, PagedResponse<T>> pageRetriever;
+        private volatile String continuationToken;
+        private volatile boolean done;
+
+        PagedByIterator(Function<String, PagedResponse<T>> pageRetriever) {
+            this.pageRetriever = pageRetriever;
+        }
+
+        @Override
+        public E next() {
+            if (!hasNext()) {
+                throw LOGGER.logThrowableAsError(new NoSuchElementException("Iterator contains no more elements."));
+            }
+
+            return getNext();
+        }
+
+        @Override
+        public boolean hasNext() {
+            // Request next pages in a loop in case we are returned empty pages for the by item implementation.
+            while (!done && needToRequestPage()) {
+                requestPage();
+            }
+
+            return isNextAvailable();
+        }
+
+        abstract boolean needToRequestPage();
+
+        abstract boolean isNextAvailable();
+
+        abstract E getNext();
+
+        synchronized void requestPage() {
+            AtomicBoolean receivedPages = new AtomicBoolean(false);
+            PagedResponse<T> page = pageRetriever.apply(continuationToken);
+            if (page != null) {
+                receivePage(receivedPages, page);
+            }
+        }
+
+        abstract void addPage(PagedResponse<T> page);
+
+        private void receivePage(AtomicBoolean receivedPages, PagedResponse<T> page) {
+            receivedPages.set(true);
+            addPage(page);
+
+            continuationToken = page.getNextLink();
+            this.done = continuationToken == null || continuationToken.isEmpty();
+        }
```

Comment:
The `receivePage` (and related var) would be different, if server support both `continuationToken` and `nextLink`.

SDK likely would only provide one `pageRetriever` with either of the approach. It may also provide a e.g. `continuationTokenRetriever` to specify whether to call `getNextLink` or `getContinuationToken`.